### ISSUE TITLE
Scalar Plot for Fitness parameter for YOLO V8

### ISSUE THREAD

**Issue by yashtiwari96** [ultralytics/ultralytics#13177]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am using clearml to log my model training. The training scalar plots includes most of  validation metrics per epoch except the "fitness" parameter. I wanted to include the "fitness" parameter as well to tune the model using clearml. Can you please help me in setting this up. I am attaching the validation scalar plot for reference
Thanks
![val](https://github.com/ultralytics/ultralytics/assets/72273439/30329db3-4b59-4b27-a2b1-124f61d672b8)

**Comment by glenn-jocher** [comment]
Hello,

Thank you for your question! To include the "fitness" parameter in your ClearML logs, you'll need to ensure that this metric is being logged during your training process. The "fitness" parameter is typically a composite metric used to evaluate the overall performance of the model.

Here's a general approach to log the "fitness" parameter:

1. **Modify Training Script**: Ensure your training script calculates and logs the "fitness" metric. You can do this by adding a custom logging function that logs the "fitness" value to ClearML.

2. **ClearML Integration**: If you haven't already, integrate ClearML into your training script. This involves initializing a ClearML Task and using the ClearML Logger to log custom metrics.

Here's a simplified example of how you might modify your training script:

```python
from clearml import Task, Logger

# Initialize ClearML Task
task = Task.init(project_name='Your Project', task_name='YOLOv8 Training')

# Your training loop
for epoch in range(epochs):
    # Training code...
    
    # Calculate fitness
    fitness = calculate_fitness(metrics)  # Replace with your fitness calculation
    
    # Log fitness to ClearML
    Logger.current_logger().report_scalar("Fitness", "value", iteration=epoch, value=fitness)
```

3. **Fitness Calculation**: Ensure you have a function to calculate the fitness metric based on your validation metrics. This function should be called at the end of each epoch.

For more detailed guidance on hyperparameter tuning and logging, you can refer to the [Hyperparameter Tuning Guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/).

If you need further assistance, feel free to ask. Happy training! 🚀

**Comment by yashtiwari96** [comment]
Thank you for your support. Clearml uses the scalar parameters metrics for finding the best hyper-parameters. Since fitness was not a part of the scalar metrics, I was not able to tune my models with that parameter for my object detection models. I was able to figure out this issue. Every trainer class has fitness attribute which is updated for each epoch. I added one line in the followinf file ultralytics/utils/callbacks/clearml.py:
def on_fit_epoch_end(trainer):
    """Reports model information to logger at the end of an epoch."""
    if task := Task.current_task():
        # You should have access to the validation bboxes under jdict
        task.get_logger().report_scalar(
            title="Epoch Time", series="Epoch Time", value=trainer.epoch_time, iteration=trainer.epoch
        )
        for k, v in trainer.metrics.items():
            task.get_logger().report_scalar("val", k, v, iteration=trainer.epoch)
         ## Just added the below line
        task.get_logger().report_scalar("val","metrics/fitness",trainer.fitness,iteration=trainer.epoch)
        if trainer.epoch == 0:
            from ultralytics.utils.torch_utils import model_info_for_loggers

            for k, v in model_info_for_loggers(trainer).items():
                task.get_logger().report_single_value(k, v)
After this, it is reporting the fitness metric as well and I am able to tune the model more efficiently.
![val (1)](https://github.com/ultralytics/ultralytics/assets/72273439/0a00456a-0abf-45dd-806b-ceece2ca33b7)
Thank you for your detailed documentation and support, I was able to resolve this

**Comment by glenn-jocher** [comment]
Hello,

Thank you for sharing your solution and the detailed explanation! It's great to hear that you were able to resolve the issue and successfully log the "fitness" metric with ClearML. Your approach of modifying the `on_fit_epoch_end` callback to include the fitness metric is spot on. This will indeed help in tuning your models more efficiently.

We appreciate your kind words about our documentation and support. If you have any more questions or need further assistance, feel free to reach out. Happy tuning! 🚀
